### ISSUE TITLE
`test_normalize_url*` failures with Python 3.10.15, 3.11.10

### ISSUE THREAD

**Issue by mgorny** [sissaschool/xmlschema#416]
```pytb
$ tox -e py310
py310: install_deps> python -I -m pip install 'elementpath<5.0.0,>=4.4.0' jinja2 lxml
.pkg: install_requires> python -I -m pip install setuptools
.pkg: _optional_hooks> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_wheel> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: prepare_metadata_for_build_wheel> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_sdist> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py310: install_package_deps> python -I -m pip install 'elementpath<5.0.0,>=4.4.0'
py310: install_package> python -I -m pip install --force-reinstall --no-deps /tmp/xmlschema/../.tox/xmlschema/.tmp/package/1/xmlschema-3.4.0.tar.gz
py310: commands[0]> python -m unittest
..................................................................................................................................................................................................................................................................................................................................F...F....ssss........................................................................................................................................................................ssss.................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
======================================================================
FAIL: test_normalize_url_slashes (tests.test_locations.TestLocations)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/xmlschema/tests/test_locations.py", line 322, in test_normalize_url_slashes
    self.assertRegex(normalize_url('////root/dir1/schema.xsd'),
AssertionError: Regex didn't match: 'file:////root/dir1/schema.xsd' not found in 'file:///root/dir1/schema.xsd'

======================================================================
FAIL: test_normalize_url_with_base_unc_path (tests.test_locations.TestLocations)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/xmlschema/tests/test_locations.py", line 285, in test_normalize_url_with_base_unc_path
    self.assertEqual(url, 'file:////filer01/MY_HOME/dev/XMLSCHEMA/test.xsd')
AssertionError: 'file:///filer01/MY_HOME/dev/XMLSCHEMA/test.xsd' != 'file:////filer01/MY_HOME/dev/XMLSCHEMA/test.xsd'
- file:///filer01/MY_HOME/dev/XMLSCHEMA/test.xsd
+ file:////filer01/MY_HOME/dev/XMLSCHEMA/test.xsd
?      +


----------------------------------------------------------------------
Ran 1388 tests in 29.941s

FAILED (failures=2, skipped=8)
py310: exit 1 (30.67 seconds) /tmp/xmlschema> python -m unittest pid=230473
  py310: FAIL code 1 (40.38=setup[9.71]+cmd[30.67] seconds)
  evaluation failed :( (40.49 seconds)
```

I'm getting the same failure with CPython 3.10.15 and 3.11.10, but not PyPy3.10 7.3.17 — so I suspect it's one of the security backports in the latest round.

**Comment by brunato** [comment]
Yes, it could be `urlunsplit()` related, I'm going to check that ... thank you

**Comment by mgorny** [comment]
And I've just confirmed that downgrading to 3.10.14 makes the test pass. So yeah, python/cpython@2fa5d706843476cf43ff3b431e3c4a8c34737413 seems like a good candidate.

**Comment by s-t-e-v-e-n-k** [comment]
This probably also impacts earlier releases that got a bump at the same time, like 3.8.20 and 3.9.20, but I've not checked.

**Comment by mgorny** [comment]
> This probably also impacts earlier releases that got a bump at the same time, like 3.8.20 and 3.9.20, but I've not checked.

Yes, indeed both versions fail too.

That said, 91df0613fbf1e07d4109bd16c5cfc6e3c3da2c50 still fails on all versions prior to 3.12:

```pytb
======================================================================
FAIL: test_normalize_url_slashes (tests.test_locations.TestLocations.test_normalize_url_slashes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/xmlschema/tests/test_locations.py", line 338, in test_normalize_url_slashes
    self.assertRegex(normalize_url('////root/dir1/schema.xsd'),
AssertionError: Regex didn't match: 'file://////root/dir1/schema.xsd' not found in 'file:///tmp/xmlschema/%5C/root/dir1/schema.xsd'

======================================================================
FAIL: test_normalize_url_with_base_unc_path (tests.test_locations.TestLocations.test_normalize_url_with_base_unc_path)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/xmlschema/tests/test_locations.py", line 298, in test_normalize_url_with_base_unc_path
    self.assertEqual(url, 'file://////filer01/MY_HOME/dev/XMLSCHEMA/test.xsd')
AssertionError: 'file:///filer01/MY_HOME/dev/XMLSCHEMA/test.xsd' != 'file://////filer01/MY_HOME/dev/XMLSCHEMA/test.xsd'
- file:///filer01/MY_HOME/dev/XMLSCHEMA/test.xsd
+ file://////filer01/MY_HOME/dev/XMLSCHEMA/test.xsd
?      +++
```

So I'm guessing you need to account for three possible values (sigh):

1. `////` for pre-fix Pythons
2. `///` for fixed < 3.12
3. `//////` for >= 3.12

That said, at this point I'm wondering if there isn't another bug in CPython to fix…

**Comment by brunato** [comment]
Ok, if it's not a problem i remove source tag 3.4.2 until I've found a resolution to this.

When the `urlunparse()` fix has been published with Python 3.12.4 and then in Py13-beta (beta3 i remember) I thought that a backport of the patch will not be applied also to Python < 3.12, to avoid possible impacts on software that use urlsplit() and urlunsplit(). So I put a rule on 3.12.4 ... but I was wrong. 

thank you

**Comment by mgorny** [comment]
I'm investigating this right now. Hopefully, soon I should know *what* is the exact difference between unpatched and patched versions of 3.12+ and earlier.

**Comment by mgorny** [comment]
Ok, so I'm seeing two differences:

1. In patched versions (3.12.4+, 3.11.10, etc.):
   `urlunsplit(("", "", r"//filer01/MY_HOME/", "", ""))`
   changed from:
   `//filer01/MY_HOME/`
   to:
   `////filer01/MY_HOME/` 
2. in 3.12+:
    `PureWindowsPath("//foo")`
   changed from:
    `\foo`
   to:
    `\\foo`

So both changes combined make `PureWindowsPath(urlunsplit(("", "", r"//filer01/MY_HOME/", "", "")))` yield:

< 3.12 unpatched: `\\filer01\MY_HOME\`
< 3.12 patched: `\filer01\MY_HOME`
3.12+ unpatched: `\\filer01\MY_HOME\`
3.12+ patched: `\\\\filer01\MY_HOME`

<details>
<summary>Script used to reproduce this</summary>

```python
import os
import os.path

from pathlib import PureWindowsPath
from unittest.mock import patch
from urllib.parse import urlsplit, quote_from_bytes, unquote, urlunsplit

path = r"dev\XMLSCHEMA\test.xsd"
base_url = PureWindowsPath("\\\\filer01\\MY_HOME\\").as_uri()
base_url_host_in_path = base_url.replace('file://', 'file:////')

url_parts = urlsplit(path)
base_url_parts = urlsplit(base_url_host_in_path)
base_path = PureWindowsPath(unquote(urlunsplit(("", base_url_parts.netloc, base_url_parts.path, "", ""))))

path = base_path.joinpath(path)
norm_path = PureWindowsPath(os.path.normpath(str(path)))
norm_uri = "file://" + quote_from_bytes(os.fsencode(norm_path.as_posix()))

print(PureWindowsPath(urlunsplit(("", "", r"//filer01/MY_HOME/", "", ""))))
```
</details>

**Comment by brunato** [comment]
Another patch in 3.12.6:

  https://github.com/python/cpython/issues/85110
  https://github.com/python/cpython/commits/3.12/Lib/urllib/parse.py

I'll try to get the point adding other tests  taken from the changes.

**Comment by mgorny** [comment]
Oh, and at this point I'm not sure whether test results need to change, or perhaps `LocationPath` should be adjusted to produce the same paths on all Python versions.

**Comment by brunato** [comment]
I agree, I'm rewriting the `from_uri()` constructor and adding a pair of helper functions.

The problem is related also to `ntpath.splitdrive()`, that has been changed a lot in the bugfix releases of Python 3.12. The current version is more precise to detect Windows drives and UNC paths.

**Comment by MeggyCal** [comment]
I confirm that https://github.com/sissaschool/xmlschema/commit/059fd3bfb305809d5ebdede1bd68dff9dbaf3348 fixes the failing test also on 3.10.15.

**Comment by mgorny** [comment]
Yeah, I can confirm that on top of 059fd3bfb305809d5ebdede1bd68dff9dbaf3348, both pre-patch and post-patch Python versions pass all tests.

**Comment by brunato** [comment]
Thank you to both for confirmation.

I was confident that all is ok now because i had simply removed the round-trip tests on `urlsplit()`, so I had already pushed the files on PyPI. 

Anyway the failing test on `urlsplit()` involve a misalignment on what urlsplit/urlparse consider an UNC path and what is considered by PureWindowsPath. The internals of PurePath had been rewritten in 3.12 so il the parsing of UNC path by urlsplit and PureWindowsPath is now coherent.

For `xmlschema` for the moment I've replaced `urlunsplit()` with a couple of function (`get_uri()` and `get_uri_path()`). I will consider to keep both or only the 2nd in v4.0 if the `urlsplit()` round-trip remain coherent in the next bugfix updates.
